runtime/sema/type_tags.go

```diff
@@ -0,0 +1,708 @@
+ * limitations under the License.
+ */
+
+package sema
+
+import (
+	"fmt"
+)
+
+// TypeTag is a bitmask representation for types.
+// Each type has a unique dedicated bit/bit-pattern in the bitmask.
+// The mask consist of two sections: `lowerMask` and the `upperMask`.
+// Each section can represent 64-types.
+//
+type TypeTag struct {
+	lowerMask uint64
+	upperMask uint64
+}
+
+var allTypeTags = map[TypeTag]bool{}
+var allLowerMaskedTypeTags []TypeTag
+var allUpperMaskedTypeTags []TypeTag
+
+func newTypeTagFromLowerMask(mask uint64) TypeTag {
+	typeTag := TypeTag{
+		lowerMask: mask,
+		upperMask: 0,
+	}
+
+	if _, ok := allTypeTags[typeTag]; ok {
+		panic(fmt.Errorf("duplicate type tag: %v", typeTag))
+	}
+
+	allTypeTags[typeTag] = true
+
+	allLowerMaskedTypeTags = append(allLowerMaskedTypeTags, typeTag)
+
+	return typeTag
+}
+
+func newTypeTagFromUpperMask(mask uint64) TypeTag {
+	typeTag := TypeTag{
+		lowerMask: 0,
+		upperMask: mask,
+	}
+
+	if _, ok := allTypeTags[typeTag]; ok {
+		panic(fmt.Errorf("duplicate type tag: %v", typeTag))
+	}
+
+	allTypeTags[typeTag] = true
+
+	allUpperMaskedTypeTags = append(allUpperMaskedTypeTags, typeTag)
+
+	return typeTag
+}
+
+func (t TypeTag) Equals(tag TypeTag) bool {
+	return t.lowerMask == tag.lowerMask &&
+		t.upperMask == tag.upperMask
+}
+
+func (t TypeTag) And(tag TypeTag) TypeTag {
+	return TypeTag{
+		lowerMask: t.lowerMask & tag.lowerMask,
+		upperMask: t.upperMask & tag.upperMask,
+	}
+}
+
+func (t TypeTag) Or(tag TypeTag) TypeTag {
+	return TypeTag{
+		lowerMask: t.lowerMask | tag.lowerMask,
+		upperMask: t.upperMask | tag.upperMask,
+	}
+}
+
+func (t TypeTag) Not() TypeTag {
+	return TypeTag{
+		lowerMask: ^t.lowerMask,
+		upperMask: ^t.upperMask,
+	}
+}
+
+func (t TypeTag) ContainsAny(typeTags ...TypeTag) bool {
+	for _, tag := range typeTags {
+		if t.And(tag).Equals(tag) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (t TypeTag) BelongsTo(typeTag TypeTag) bool {
+	return t.And(typeTag).Equals(t)
+}
+
+const noTypeMask = 0
+
+// Lower mask types
+const (
+	numberTypeMask uint64 = 1 << iota
+	signedNumberTypeMask
+	integerTypeMask
+	signedIntegerTypeMask
+	unsignedIntegerTypeMask
+	fixedPointTypeMask
+	signedFixedPointTypeMask
+	unsignedFixedPointTypeMask
+
+	intTypeMask
+	int8TypeMask
+	int16TypeMask
+	int32TypeMask
+	int64TypeMask
+	int128TypeMask
+	int256TypeMask
+
+	uintTypeMask
+	uint8TypeMask
+	uint16TypeMask
+	uint32TypeMask
+	uint64TypeMask
+	uint128TypeMask
+	uint256TypeMask
+
+	word8TypeMask
+	word16TypeMask
+	word32TypeMask
+	word64TypeMask
+
+	_ // future: Fix8
+	_ // future: Fix16
+	_ // future: Fix32
+	fix64TypeMask
+	_ // future: Fix128
+	_ // future: Fix256
+
+	_ // future: UFix8
+	_ // future: UFix16
+	_ // future: UFix32
+	ufix64TypeMask
+	_ // future: UFix128
+	_ // future: UFix256
+
+	stringTypeMask
+	characterTypeMask
+	boolTypeMask
+	nilTypeMask
+	voidTypeMask
+	addressTypeMask
+	metaTypeMask
+	blockTypeMask
```

Comment:
That was a dilemma I had :) whether to treat them like structs, or special case them, because they are 'structs', as well as pre-known. Eventually went with the first option (treating them as any other structs) so that it would behave identically to others (e.g: finding common interfaces, etc. would work seamlessly, if we add conformances someday)

I'm not sure, maybe we don't need to worry about that far into the future. I don't have a strong opinion towards either option.

Reply:
Ah, OK. Maybe it makes sense to remove some of the tags for some of the tags, like for `Block`?

When should a type tag be added/defined? Is it necessary in certain cases, or is it optional, and what are the benefits?

Reply:
So the rule of thumb for adding a type tag is: Add a unique type tag for each simple/numeric type. 
For others, add a single common type tag for each _category_ of 'derived types' (types that relies on other types) - e.g: one for composite types, one for dictionaries, etc.

In the `Block` case, the `BlockType` is actually defined as a `SimpleType` rather than as a `CompositeType`. So added a tag similar to other simple types.

Reply:
For simple/numeric types, it is required to have a unique tag. I have added a test case that ensures this, so we won't miss out on adding one.

For others, it's optional to have a unique tag (but requires a one for their 'category'). Having a unique tag for a derived type `T` would only give some performance optimization when finding the supertype of a collection of all `T`s. Because it will early exit by checking the bit of the bitmask, and don't need to fall back on checking type's deep-equality.

Reply:
Sounds good, thank you for the details @SupunS! 👍 

Could you please add these comments as comments to the code for future reference? This will be really useful when updating/extending this code